test/Table_Tests/src/Database/Upload_Spec.enso

```diff
@@ -205,6 +217,55 @@ spec make_new_connection prefix persistent_connector=True =
                 r1.should_fail_with Illegal_Argument
 
+        Test.specify "should not issue a DELETE statement for the original table name in dry run mode, even if the table does not exist" <|
+            original_table_name = Name_Generator.random_name "no-delete-test"
+
+            log_file = enso_project.data / "transient" / "sql.log"
+            log_file.delete_if_exists
+            Test_Environment.unsafe_with_environment_override "ENSO_SQL_LOG_PATH" log_file.absolute.path <|
+                Context.Output.with_disabled <|
+                    t1 = connection.create_table original_table_name structure=[Column_Description.Value "X" Value_Type.Integer] temporary=True
+                    t1.column_names . should_equal ["X"]
+
+            logs = log_file.read Plain_Text . lines
+            deletes = logs.filter (_.contains "DROP")
+            wrapped_name = connection.dialect.wrap_identifier original_table_name
+            deletes.each line->
+                if line.contains wrapped_name then
+                    Test.fail "The log file contains a dangerous DELETE statement for the original table: "+line
+
+            log_file.delete_if_exists
+
+        Test.specify "should re-use the same temporary table name across dry runs but re-create it each time" <|
+            base_table_name = Name_Generator.random_name "multi-dry-create"
+
+            Context.Output.with_disabled <|
+                t1 = connection.create_table base_table_name structure=[Column_Description.Value "X" Value_Type.Integer] temporary=True
+                Problems.expect_only_warning Dry_Run_Operation t1
+
+                tmp_name = t1.name
+                tmp_name . should_not_equal base_table_name
+                connection.query base_table_name . should_fail_with Table_Not_Found
+
+                t2 = connection.create_table base_table_name structure=[Column_Description.Value "Y" Value_Type.Char] temporary=True
+                Problems.expect_only_warning Dry_Run_Operation t2
+                t2.name . should_equal tmp_name
+                t2.at "Y" . to_vector . should_equal []
+
+                Context.Output.with_enabled <|
+                    Problems.assume_no_problems <| t2.remove_warnings.update_rows (Table.new [["Y", ['a']]])
+                t2.at "Y" . to_vector . should_equal ['a']
+
+                t3 = connection.create_table base_table_name structure=[Column_Description.Value "Y" Value_Type.Char] temporary=True
+                Problems.expect_only_warning Dry_Run_Operation t3
+                t3.name . should_equal tmp_name
+                t3.at "Y" . to_vector . should_equal []
+                # Old reference may work because it has the same structure - but contents have been rewritten:
+                t2.at "Y" . to_vector . should_equal []
+
+                # The old dry run table will not work due to different structure:
+                t1.at "X" . to_vector . should_fail_with SQL_Error
+
```

Comment:
Thanks @GregoryTravis for the comment, I've added a new test case that shows multiple dry runs for the same table. This shows the motivation for the DROP that is in there:

If I were to remove the tmp drop:
```diff
                 effective_table_name = if dry_run.not then table_name else table_naming_helper.generate_dry_run_table_name table_name
                 effective_temporary = temporary || dry_run
                 created_table_name = Context.Output.with_enabled <|
-                    if dry_run then
-                        ## This temporary table can be safely dropped if it
-                           exists, because it only existed if it was created by
-                           a previous dry run. `generate_dry_run_table_name`
-                           will never return a name of a table that exists but
-                           was created outside of a dry run.
-                        connection.drop_table effective_table_name if_exists=True
                     internal_create_table_structure connection effective_table_name structure primary_key effective_temporary on_problems
                 if dry_run.not then connection.query (SQL_Query.Table_Name created_table_name) else
                     created_table = connection.base_connection.internal_allocate_dry_run_table created_table_name
```

then that test would fail with
```
[SQLite] Creating an empty table:  [16/17, 4245ms]
    - [FAILED] should re-use the same temporary table name across dry runs but re-create it each time [145ms]
        Reason: Expected only warning Dry_Run_Operation, but a dataflow error has been matched: There was an SQL error: [SQLITE_ERROR] SQL error or missing database (table "enso-dry-run-multi-dry-create-c41774f6-74d4-4cb8-a947-0f4e027209a3" already exists). [Query was: CREATE TEMPORARY TABLE "enso-dry-run-multi-dry-create-c41774f6-74d4-4cb8-a947- …] (at C:\NBO\enso\test\Table_Tests\src\Database\Upload_Spec.enso:251:17-65).
```